opbeat.go

```diff
@@ -0,0 +1,480 @@
+	if r.TLS != nil {
+		scheme = "https"
+	} else {
+		scheme = "http"
+	}
+
+	http := HTTP{
+		scheme + "://" + r.Host + r.URL.String(),
+		r.Method,
+		headers,
+	}
+
+	if options == nil {
+		options = new(Options)
+	}
+
+	options.HTTP = &http
+
+	return opbeat.CaptureErrorSkip(e, 4, options)
+}
+
+// Captures an error and skips `skip` amount of frames in the stacktrace, this
+// is useful to stop logging library type frames to Opbeat. Also takes a map of
+// other interfaces that are written to Opbeat as a part of the log. Please take
+// care that any values in this map can be marshalled into JSON.
+func (opbeat *Opbeat) CaptureErrorSkip(e error, skip int, options *Options) error {
+	stacktrace, err := stacko.NewStacktrace(3 + skip)
+	if err != nil {
+		return err
+	}
+
+	p, err := newPacket(e.Error(), stacktrace, options)
+	if err != nil {
+		return err
+	}
+
+	p.Revision = opbeat.Revision
+	p.Level = Error
+	p.Logger = opbeat.LoggerName
+
+	opbeat.queue(p)
+
+	return nil
+}
+
+// Captures a message along with a level indicating the severity of the message.
+func (opbeat *Opbeat) CaptureMessage(message string, l Level, options *Options) error {
+	p, err := newPacket(message, nil, options)
+	if err != nil {
+		return err
+	}
+
+	p.Level = l
+
+	opbeat.queue(p)
+
+	return nil
+}
+
+// Waits for all packets to send.
+func (opbeat *Opbeat) Wait() {
+	opbeat.wait.Wait()
+}
+
+// Starts a goroutine which listens on the main channel and sends any packets it
+// receives to Opbeat. The goroutine will exit if the client's `.Close` function
+// is called.
+func (opbeat *Opbeat) start() {
+	opbeat.packets = make(chan *packet, 20)
+	go func() {
+		var p *packet
+		var open bool
+		for {
+			select {
+			case p, open = <-opbeat.packets:
+				if !open {
+					return
+				}
+				err := opbeat.send(p)
+				if err != nil {
+					opbeat.Logger.Println(err)
+				}
+				opbeat.wait.Done()
+			}
+		}
+	}()
+}
+
+// Waits for all requests to finish and closes the main channel. Closing the
+// channel will force the goroutines communicating packets to return. This
+// effectively kills the client and a new will have to be created to continue
+// communicating with Opbeat.
+func (opbeat *Opbeat) Close() {
+	opbeat.Wait()
+	close(opbeat.packets)
+}
+
+func (opbeat *Opbeat) queue(p *packet) {
+	opbeat.wait.Add(1)
+	opbeat.packets <- p
+}
+
+func (opbeat *Opbeat) send(p *packet) error {
+	body, err := json.Marshal(p)
+	if err != nil {
+		return err
+	}
+
+	url := fmt.Sprintf("https://%s/api/v1/organizations/%s/apps/%s/errors/",
+		opbeat.Host, opbeat.organizationId, opbeat.appId)
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(body))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Authorization", "Bearer "+opbeat.secretToken)
+
+	res, err := opbeat.Client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	body, err = ioutil.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	res.Body.Close()
```

Comment:
@birkirb recommends moving this up to L322 and using `defer` to make sure we close even when we get an error :+1: